### ISSUE TITLE
Improve QmlWeb's memory management

### ISSUE THREAD

**Issue by akreuzkamp** [qmlweb/qmlweb#150]
in QML you can delete Items. Think e.g. of a Repeater, whose model-number decreases. We need to make sure, that all bindings and references are deleted and the object is garbage-collected.

Coming from the C++ world, I have no clue how to do proper memory management in garbage-collected languages. So far, the ideas I have are:
1) Create a big array of all objects we have. There we'd have a reference to each object, everywhere else we'd not hold references, but only the index of the object in our big array. Dereferencing would be bigArray[someBindingId]. Now we can delete objects by simply resetting the array entry to null. As soon as we encounter a reference to a deleted object, we'll recognize and clean up. Performance impact would probably be low (array-access is very fast), but we'd have dead references lying around and would need a lot of boilerplate code to check if a reference is still valid.
2) Keep lists like "$tidyupList" containing references to child objects, bindings and stuff we need to tidy up when deleting objects. This is how it works currently. It needs no boilerplate code, but is very weak-kneed, as it's easy to miss a reference.

I'm not really happy with either way...
I hope you've got some good input! :)

**Comment by akreuzkamp** [comment]
Option 3) A mix of both: We could try to rebuild Qt's pointer classes in js, especially QPointer and QWeakPointer.

* QPointer is a pointer that resets itself as soon as the object it points to is deleted. We could achieve that by making a QPointer be a function object that would hold and return our reference. The object it points to would have list of all QPointers pointing to it and would reset all of them from its destructor. That's basically (2) done more nicely.
* QWeakPointer is a pointer that doesn't prevent garbage collecting (in Qt it only makes sense in conjunction with QSharedPointer, which is a reference-counted pointer and thus (almost) behaves like js-references ;) ). We'd just implement it like I proposed with (1).

We should then always use one of those two, trying to avoid direct references.

**Comment by akreuzkamp** [comment]
To clarify: One of the concerns of this issue is to ask you, how to properly do memory management in JavaScript ;)

**Comment by labsin** [comment]
If an object is unreachable, it should get freed when the garbage collector
kicks in. It starts at the root and marks every object it can reach and
deletes the unreachable. There is no (cross browser) way to start a new
run. It's also unclear and very implementation specific when it kicks in.
Removing every reference is all you can do. Cycle references (a references
b and b a) shouldn't be an issue if both are unreachable.
The problem is the same as with shared pointers. See that there is no
reference left. Maybe there is a problem with qml signals or properties.
Op 23 mrt. 2016 20:08 schreef "akreuzkamp" <notifications@github.com>:

**Comment by akreuzkamp** [comment]
Thanks for your reply, @labsin. I know the general concept of garbage collecting, though. The idea behind gc is, that you don't need to take care of memory management. But in QmlWeb we need the possibility to explicitly delete elements.

In C++ that means: We simply delete the object. If we still have pointers to that object, they stay as they are. As soon as we try to dereference the dangling pointer, we'll crash. The concequence: We need to properly remove all dangling pointers.
In JS that means: We need to make the garbage collector clean up the object. The obvious consequence is the same as in C++: We need to properly remove all dangling references.

The question is, *how* to properly remove all dangling references. Are there any established mechanisms for that in JavaScript?

**Comment by travellhyne** [comment]
There's WeakMap and WeakSet. But these things are ES2015 specs which means that they are not widely supported among browsers yet.

**Comment by labsin** [comment]
I don't think weakmap and weakset can be used for this. There is no way to
iterate over a weakmap or weakset.
The array and keeping the index only could be used, but then we need to
check if it still exists every time. I also think it's a performance hit.
I think the biggest problem is if there is a reference to objects in a
closure somewhere. They are not that obvious.

The weakset can only be used to "tag" an item (is it processed yet).
Weakmap can be used if you have a library (like this one) that returns
objects with some function. Now if you need to store info about this object
in the library that you don't want to expose in the object, you can store
it in a weakmap so that there is no hanging reference after the user of the
library deletes it.
If the user then passes the object to some other function of the library,
it can get the needed info from the map.

The weakmap could be handy too keep track of binding loops. Too mark it and
only stop on the 3th loop and return the same value in stead of expanding
the binding again (like in qt).
Op 25 mrt. 2016 15:26 schreef "Neil John Ramal" <notifications@github.com>:

**Comment by henrikrudstrom** [comment]
Is this really an issue? As long as we dont maintain any global list of references to objects somewhere, the javascript garbage collector should clean up any objects and their children once the object is no longer referenced anywhere.

**Comment by akreuzkamp** [comment]
We maintain lot's of references to items at random locations, but still we want to explicitly delete objects. 

The gc won't clean up anything, if we don't remove all references, which we nearly can't, because we don't know all the locations.

This *is* an issue, because those references *will* be used and they will fail, because our object is supposed to be deleted.

**Comment by jhihn** [comment]
QML has gc() http://doc.qt.io/qt-5/qtqml-javascript-qmlglobalobject.html But it is not standard. Why maintain a reference to something that that's not needed? If you're looking beyond GC to relocations, then that's not really anything you can control. It's up to the browser to invoke the GC. You're trying to to too much. It's just a webpage after all. If it's not needed then release it and let the browser do its thing.

**Comment by akreuzkamp** [comment]
I'm not mainly arguing about memory usage.
In C++ having a dangling reference will crash, in JS having a dangling reference will simply produce undefined behavior. That's what I'm arguing.

> Why maintain a reference to something that that's not needed?
Obviously, you should not. But it's not that simple®. If you have just one reference, it's easy. Release it. But tell me one example of a real application, where each object is only referenced once. I'm arguing that we need to find a way to release objects, properly. To release every reference that exists.

**Comment by ChALkeR** [comment]
@akreuzkamp What do you mean by a «dangling reference» and «undefined behavior»? Could you provide some code example?

**Comment by akreuzkamp** [comment]
@ChALkeR With "dangling reference" I mean a reference to an object, that is supposed to be deleted. e.g.
```qml
FocusScope {
    id: fs
    property Item fi
    Repeater {
        id: rep
        model: 5

        Rectangle {
            width: 50; height: 50; color: "orange"
            onActiveFocusChanged: if (rep.model == 0) console.log("I'm wrong behavior");
        }
    }

    Component.onCompleted: {
        fi = rep.itemAt(3);
        fi.forceActiveFocus();
    }
}
```
`fs` now holds a reference to the 4th item of the repeater. What will happen to it, when `rep.model` is set to 0? Now it's a "dangling reference". Obviously it's not dangling in the C++ way, as the object is not destructed yet. But still it's logically deleted.

With "undefined behavior" I mean whatever happens if you call a function on an object, that is supposed not to exist anymore. E.g. if `fs` got/lost focus and thus sets `activeFocus` on `fi`. That could trigger wrong console outputs, exceptions, wrong data to be sent to the server or Great Britain to leave the EU.

**Comment by stephenmdangelo** [comment]
@akreuzkamp 

> or Great Britain to leave the EU

Ha, that made me lol.

**Comment by jhihn** [comment]
I see what you mean, however you should also have a onModelChanged event handler that clears any potential invalid references. In your example, it should null out `fi`. I don't think that's an QML engine issue, I think it's a JS GC issue, and developers in a language should know how the GC works.

**Comment by stephenmdangelo** [comment]
> I don't think that's an QML engine issue, I think it's a JS GC issue, and developers in a language should know how the GC works.

+1

**Comment by akreuzkamp** [comment]
Thanks, now we can start to argue. :)

imo, it's bad style to rely on onModelChanged/QObject.$delete() know all possible references to the object. It's easy to miss a reference and it badly hurts modularity. It will definitely fail, as soon as a type stores a reference, that isn't part of QtQuick, but is dynamically loaded from a module. How should Repeater/QObject know about it?
That leads us to the insight, that we need some structured way to manage those references, right?

**Comment by stephenmdangelo** [comment]
@akreuzkamp In the example you gave, this just seems like user-error. If I'm using QML, I need to know that Items within a Repeater (or a Loader, or anything that dynamically creates objects) will be deleted as a result of certain operations.

**Comment by jhihn** [comment]
The only way around that is to make the assignment of `fi` an indirect one. such that it is a reference to the expression `rep.itemAt(3)`. At which point you go down an rabbit hole of indirect references to indirect references, and I don't want to imagine that because you keep stacking on references on every modify operation. Consider an array, `var a = [1,2,3]`, now I can have 4 indirect references, `a`, `a[0]`, `a[1]`, `a[2]`, which you want to be evaluated every time. But If I make additional references, `var ab = [ a[0] ]` now I have two levels references. Any sane engine will give you only a reference to the concrete item. You either have to copy `a[0]`'s value at assignment to `ir` or track `ir[ a[0] ]` and then how do you the programmer indicate which you mean? In other languages it is easy. In Javascript there is no syntactic sugar available. So I say we stick to the engine's rules because that's what the engine user's expect. My understanding is you get a copy of `a[0]` with `ir[ a[0] ]` such that `delete a[0]` does not remove the item from ir. Meaning that under engine rules, your idea is invalid?

   a = [{'A':1},{'B':2}] 
=> [ { A: 1 }, { B: 2 } ]
   b= a[1]
=> { B: 2 }
   delete a[1];
=> true
   a
=> [ { A: 1 } ]
   b
=> { B: 2 }

**Comment by akreuzkamp** [comment]
@stephenmdangelo: no, it's not a user error. In reality, you wouldn't write `rep.itemAt(3).forceActiveFocus()` but the user would click on the forth item of the list. The consequence is the same: The FocusScope will hold a (internal!) reference to a dynamically created and deleted item. If that reference isn't cleared, `activeFocusChanged` (and more) will be triggered.

@jhihn I don't get your point. But I don't expect `fi` to magically point to some new object again, once the model gets repopulated. But I do expect (and this is what qt does) `fi` to be `null` after resetting the model. Still, to make it clear: I was arguing about an *internal* reference, that every FocusScope holds to the focused Item.

Let me redefine my example more precisely:
```qml
FocusScope {
    id: fs
    Repeater {
        id: rep
        model: 5

        Rectangle {
            width: 50; height: 50; color: "orange"
            onActiveFocusChanged: if (rep.model == 0) console.log("I'm wrong behavior");
        }
    }

    Component.onCompleted: {
        rep.itemAt(3).forceActiveFocus(); // to be called fi
    }
}
```

**Comment by jhihn** [comment]
Yes, I gathered that. If that's what you mean, then write it. However hardcoding indices is asking for trouble (as I am sure you are aware). In the meantime, itemAt() it will return null when model==0, 
`itemAt(index) - Returns the Item that has been created at the given index, or null if no item exists at index.`
and produce the familiar cannot call a function of null error. I believe this is why Qt provides their own index operations of get() and itemAt(), etc. I *think* it let's them work around the GC issues?

**Comment by akreuzkamp** [comment]
Sorry, I could have made it clearer. But again, it seems you missed my point. :(
Replace the `rep.itemAt(3).forceActiveFocus();` by the user giving focus to one item of the repeater, by clicking on it or using Tab or whatever. I only want to point out a common situation, where the FocusScope will hold a reference to a dynamically created and deleted object, no matter how it gets it.

**Comment by stephenmdangelo** [comment]
> Replace the rep.itemAt(3).forceActiveFocus(); by the user giving focus to one item of the repeater, by clicking on it or using Tab or whatever.

But the Repeater would have already called `$delete`, which I believe removes the item from the DOM, so interacting with it wouldn't be an option.

> Sorry, I could have made it clearer. But again, it seems you missed my point. :(

I'm not trying to focus on this specific example, but is there some other sort of example that you can see leading to a problem that wouldn't also be a problem in C++ QML?  Or is your point that in C++ QML there is defined behaviour for accessing a delete Object, whereas in qmlweb, you still have a perfectly valid reference to an object (despite it being half "deleted" in the sense that it has been removed from the DOM and qmlweb Object tree)?

In regards to your earlier comment about a crash:

> In C++ that means: We simply delete the object. If we still have pointers to that object, they stay as they are. As soon as we try to dereference the dangling pointer, we'll crash. The concequence: We need to properly remove all dangling pointers.

The C++ QML Engine protects against crashing when accessing deleted QML-created objects (like in your Repeater example).  "fi" would end up just being "null".

**Comment by akreuzkamp** [comment]
> But the Repeater would have already called $delete, which I believe removes the item from the DOM, so interacting with it wouldn't be an option.

No. You click or use tab to focus the item. *Now* you clear the model. Now fs.$focusItem still is a valid reference to the deleted item.

> I'm not trying to focus on this specific example, but is there some other sort of example that you can see leading to a problem that wouldn't also be a problem in C++ QML?

I fixed all errors, that occurred with it, but not in a way, that makes me happy. I've seen the items' parent calling some functions on the deleted child, which caused the child to try to access its deleted dom. I think we now have special handling for that, but I don't like it.

> The C++ QML Engine protects against crashing when accessing deleted QML-created objects (like in your Repeater example). "fi" would end up just being "null".

Yes, it uses QPointer to do that (see Option 3, above).

**Comment by stephenmdangelo** [comment]
> fs.$focusItem still is a valid reference to the deleted item.

Ah ok, I wasn't understanding that your example was referring to an internal QObject reference stored within FocusScope. Obviously builtin qmlweb components shouldn't be referring to deleted QObjects, just as a user of qmlweb shouldn't. 

For QObjects stored in a QMLProperty, we could check in "get" if the QObject has been deleted and return null to simulate C++ QML behaviour. Don't know if it's worth the (maybe neglible) performance impact?

Adding a "destroyed" signal emitted in $delete, as C++ QObject does, could be connected to by anything holding a reference to a non-owned QObject (and could be used to implement a QPointer-type class for convenience).
